ext/ldap/ldap.c

```diff
@@ -2491,351 +2491,255 @@ PHP_FUNCTION(ldap_delete_ext)
-		_ldap_hash_fetch(mod, LDAP_MODIFY_BATCH_VALUES, &vals);
+		zval *attrib = zend_hash_str_find(Z_ARRVAL_P(modification_zv), LDAP_MODIFY_BATCH_ATTRIB, strlen(LDAP_MODIFY_BATCH_ATTRIB));
+		ZEND_ASSERT(Z_TYPE_P(attrib) == IS_STRING);
+		zval *modtype = zend_hash_str_find(Z_ARRVAL_P(modification_zv), LDAP_MODIFY_BATCH_MODTYPE, strlen(LDAP_MODIFY_BATCH_MODTYPE));
+		ZEND_ASSERT(Z_TYPE_P(modtype) == IS_LONG);
+		zval *modification_values = zend_hash_str_find(Z_ARRVAL_P(modification_zv), LDAP_MODIFY_BATCH_VALUES, strlen(LDAP_MODIFY_BATCH_VALUES));
+		ZEND_ASSERT(modification_values == NULL || Z_TYPE_P(modification_values) == IS_ARRAY);
 
 		/* map the modification type */
+		int ldap_operation;
 		switch (Z_LVAL_P(modtype)) {
 			case LDAP_MODIFY_BATCH_ADD:
-				oper = LDAP_MOD_ADD;
+				ldap_operation = LDAP_MOD_ADD;
 				break;
 			case LDAP_MODIFY_BATCH_REMOVE:
 			case LDAP_MODIFY_BATCH_REMOVE_ALL:
-				oper = LDAP_MOD_DELETE;
+				ldap_operation = LDAP_MOD_DELETE;
 				break;
 			case LDAP_MODIFY_BATCH_REPLACE:
-				oper = LDAP_MOD_REPLACE;
+				ldap_operation = LDAP_MOD_REPLACE;
 				break;
-			default:
-				zend_throw_error(NULL, "Unknown and uncaught modification type.");
-				RETVAL_FALSE;
-				efree(ldap_mods[i]);
-				num_mods = i;
-				goto cleanup;
+			EMPTY_SWITCH_DEFAULT_CASE();
 		}
 
 		/* fill in the basic info */
-		ldap_mods[i]->mod_op = oper | LDAP_MOD_BVALUES;
-		ldap_mods[i]->mod_type = estrndup(Z_STRVAL_P(attrib), Z_STRLEN_P(attrib));
+		ldap_mods[modification_index]->mod_op = ldap_operation | LDAP_MOD_BVALUES;
+		ldap_mods[modification_index]->mod_type = estrndup(Z_STRVAL_P(attrib), Z_STRLEN_P(attrib));
 
 		if (Z_LVAL_P(modtype) == LDAP_MODIFY_BATCH_REMOVE_ALL) {
 			/* no values */
-			ldap_mods[i]->mod_bvalues = NULL;
-		}
-		else {
+			ldap_mods[modification_index]->mod_bvalues = NULL;
+		} else {
 			/* allocate space for the values as part of this modification */
-			num_modvals = zend_hash_num_elements(Z_ARRVAL_P(vals));
-			ldap_mods[i]->mod_bvalues = safe_emalloc((num_modvals+1), sizeof(struct berval *), 0);
+			uint32_t num_modification_values = zend_hash_num_elements(Z_ARRVAL_P(modification_values));
+			ldap_mods[modification_index]->mod_bvalues = safe_emalloc((num_modification_values+1), sizeof(struct berval *), 0);
 
 			/* for each value */
-			for (j = 0; j < num_modvals; j++) {
-				/* fetch it */
-				fetched = zend_hash_index_find(Z_ARRVAL_P(vals), j);
-				modval = zval_get_string(fetched);
+			zend_ulong value_index = 0;
+			zval *modification_value_zv = NULL;
+			ZEND_HASH_FOREACH_NUM_KEY_VAL(Z_ARRVAL_P(modification_values), value_index, modification_value_zv) {
+				zend_string *modval = zval_get_string(modification_value_zv);
 				if (EG(exception)) {
 					RETVAL_FALSE;
-					ldap_mods[i]->mod_bvalues[j] = NULL;
-					num_mods = i + 1;
+					ldap_mods[modification_index]->mod_bvalues[value_index] = NULL;
+					num_mods = modification_index + 1;
 					goto cleanup;
 				}
```

Comment:
This case should probably be moved to the validation step ensuring that all values are strings.